arangod/Aql/QueryPlanCache.cpp

```diff
@@ -0,0 +1,356 @@
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Jan Steemann
+////////////////////////////////////////////////////////////////////////////////
+
+#include "QueryPlanCache.h"
+
+#include "Aql/QueryOptions.h"
+#include "Basics/Exceptions.h"
+#include "Basics/VelocyPackHelper.h"
+#include "Basics/conversions.h"
+#include "Basics/debugging.h"
+#include "Basics/system-functions.h"
+#include "Logger/Logger.h"
+#include "Metrics/Counter.h"
+#include "Metrics/Gauge.h"
+#include "Random/RandomGenerator.h"
+#include "VocBase/vocbase.h"
+
+#include <boost/container_hash/hash.hpp>
+#include <velocypack/Builder.h>
+#include <velocypack/Iterator.h>
+#include <velocypack/Options.h>
+#include <velocypack/Slice.h>
+
+namespace arangodb::aql {
+
+size_t QueryPlanCache::Key::hash() const {
+  return QueryPlanCache::KeyHasher{}(*this);
+}
+
+size_t QueryPlanCache::Key::memoryUsage() const noexcept {
+  // note: the 256 bytes overhead is a magic number estimated here.
+  // we will definitely have some overhead for each entry because
+  // strings may have reserved more capacity than actual bytes
+  // are used. the velocypack buffer for the bind parameters also
+  // may have some overhead because of over-allocation.
+  return 256 + queryString.size() + bindParameters->byteSize();
+}
+
+size_t QueryPlanCache::KeyHasher::operator()(
+    QueryPlanCache::Key const& key) const noexcept {
+  size_t hash = 0;
+  boost::hash_combine(hash, key.queryString.hash());
+  boost::hash_combine(hash,
+                      VPackSlice(key.bindParameters->data()).normalizedHash());
+  // arbitrary integer values used here from sha1sum-ing
+  // for fullcount=true / fullcount=false
+  // for forceOneShard=true / forceOneShard=false
+  boost::hash_combine(hash, key.fullCount ? 0x8ca49570 : 0xcd9ff1e8);
+  boost::hash_combine(hash, key.forceOneShard ? 0x36c0c09b : 0x20e96f11);
+  return hash;
+}
+
+bool QueryPlanCache::KeyEqual::operator()(
+    QueryPlanCache::Key const& lhs,
+    QueryPlanCache::Key const& rhs) const noexcept {
+  if (lhs.fullCount != rhs.fullCount) {
+    return false;
+  }
+
+  if (!lhs.queryString.equal(rhs.queryString)) {
+    return false;
+  }
+
+  auto lhsSlice = VPackSlice(lhs.bindParameters->data());
+  auto rhsSlice = VPackSlice(rhs.bindParameters->data());
+  if (lhsSlice.normalizedHash() != rhsSlice.normalizedHash()) {
+    return false;
+  }
+  if (basics::VelocyPackHelper::compare(
+          lhsSlice, rhsSlice,
+          /*useUtf8*/ true, &VPackOptions::Defaults, nullptr, nullptr) != 0) {
+    return false;
+  }
+  return true;
+}
```

Comment:
This code is about to fail tomorrow when someone extends the cache key with a new field. Instead one could use `operator==() = default` which is provided by c++20 and let the compiler automatically generate this comparison and not forget anything.

Reply:
LOL. It is already broken now as we speak, since I have added fields. Will fix. However, the C++ compiler will probably not be able to ensure proper comparison of the slices, right?